### PR TITLE
Implement substring search feature

### LINKED ISSUE #19: Substring search
As a user, I can search for specific names, tags or assets, so that I can more easily locate relevant information.

### PR DESCRIPTION
**Pull request by rizkidelta**
Implemented substring search to replace the previous keyword search feature. Search is now more forgiving with whitespaces, allowing partial matches.

Closes #19

### CODE REVIEW

File: src/main/java/seedu/address/model/person/PersonMatchesKeywordsPredicate.java [hunk 1]
```diff
@@ -20,17 +19,58 @@ public PersonMatchesKeywordsPredicate(List<String> keywords) {
...
-                           boolean assetsContainKeyword = person.getAssets()
-                                   .stream()
-                                   .anyMatch(asset -> StringUtil.containsWordIgnoreCase(asset.get(), keyword));
-                           return nameContainsKeyword || tagsContainKeyword || assetsContainKeyword;
-                       });
+                .anyMatch(keyword -> fuzzyMatch(person.getName().toString(), keyword)
+                        ||
+                        person.getTags().stream().anyMatch(tag -> fuzzyMatch(tag.get(), keyword))
+                        ||
+                        person.getAssets().stream().anyMatch(asset -> fuzzyMatch(asset.get(), keyword)));
```
Comment (aureliony) [inline src/main/java/seedu/address/model/person/PersonMatchesKeywordsPredicate.java:26]: Would it be better to place the operators on the same lines as the statements?

File: src/main/java/seedu/address/model/person/PersonMatchesKeywordsPredicate.java [hunk 2]
```diff
@@ -20,17 +19,58 @@ public PersonMatchesKeywordsPredicate(List<String> keywords) {
     @Override
     public boolean test(Person person) {
```
Comment (aureliony) [inline src/main/java/seedu/address/model/person/PersonMatchesKeywordsPredicate.java:20]: Please update the testcases to test the fuzzy search.

**Review by aureliony — CHANGES_REQUESTED** [review]

**Review by aureliony — APPROVED** [review]

File: src/main/java/seedu/address/logic/commands/FindCommand.java [hunk 3]
```diff
@@ -3,30 +3,31 @@
...
 import seedu.address.logic.Messages;
 import seedu.address.model.Model;
-import seedu.address.model.person.PersonMatchesKeywordsPredicate;
+import seedu.address.model.person.Person;
+import seedu.address.model.person.PersonMatchesQueryPredicate;
 
 /**
- * Finds and lists all persons in address book whose name contains any of the argument keywords.
- * Keyword matching is case insensitive.
+ * Finds and lists all persons in the address book such that certain fields match the predicate,
```
Comment (yisiox) [inline src/main/java/seedu/address/logic/commands/FindCommand.java:14]: Nit: should this end with a period?

File: src/main/java/seedu/address/model/person/PersonMatchesQueryPredicate.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+/**
+ * Tests that a {@code Person}'s {@code Name}, {@code Tags} or {@code Assets} matches the search query.
+ * The algorithm removes all whitespaces and checks if the query is a substring of a field.
+ * The substring matching is case-insensitive.
+ */
+public class PersonMatchesQueryPredicate implements Predicate<Person> {
+
+    private static final String whitespaceRegex = "\\s+";
+    private static final String emptyString = "";
```
Comment (yisiox) [inline src/main/java/seedu/address/model/person/PersonMatchesQueryPredicate.java:15]: Not sure but these seem like constants. Should it be in SCREAMING_SNAKE_CASE?

**Review by yisiox — COMMENTED** [review]
Looks good overall, perhaps check two nits.

File: src/main/java/seedu/address/logic/commands/FindCommand.java (hunk 3 above)
Comment (aureliony) [reply src/main/java/seedu/address/logic/commands/FindCommand.java:14]: Yes, fixed.

File: src/main/java/seedu/address/model/person/PersonMatchesQueryPredicate.java (hunk 4 above)
Comment (aureliony) [reply src/main/java/seedu/address/model/person/PersonMatchesQueryPredicate.java:15]: Yes, changed.

**Pull request merged**